Graduate Backoff Limit Per Index as stable

mimowo opened the pull request:
#### What type of PR is this?

/kind feature
/kind documentation

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:

Enhancement issue: https://github.com/kubernetes/enhancements/issues/3850

#### Special notes for your reviewer:

This PR proposes to promote the e2e tests as conformance too, here is the summary: https://github.com/kubernetes/kubernetes/pull/130061#issuecomment-2662525387.

I'm also ok to separate it out, but to me combining makes sense, as promotion of the e2e tests was one of the graduation criteria for stable.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Promote the Job's JobBackoffLimitPerIndex feature-gate to stable.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs
```

mimowo commented:
cc @atiratree @soltysh @kannon92

mimowo committed:
Graduate Backoff Limit Per Index as stable

Reenable the JobBackoffLimitPerIndex_Reenabling integration test

tenzen-y commented:
Let me try to take this as a first review.
/cc

tenzen-y commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -2437,6 +2437,10 @@ func TestTrackJobStatusAndRemoveFinalizers(t *testing.T) {
+				// TODO: this will be removed in 1.36
+				featuregatetesting.SetFeatureGateEmulationVersionDuringTest(t, feature.DefaultFeatureGate, utilversion.MustParse("1.32"))
+			}
 			featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobBackoffLimitPerIndex, tc.enableJobBackoffLimitPerIndex)
```
```suggestion
```

Do we still need to perform this?
IIUC, we can delegate this to the above emulation, right?

mimowo replied:
Yes, commenting out this line makes the tests fail. 

An alternative which works is moving this line under the `if` - next to the `SetFeatureGateEmulationVersionDuringTest` call. 

However, moving this line generates extra diff which is not necessary, so maybe better not to. Also, I followed the pattern for other feature-gates, like this one: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control_test.go#L223-L227.

tenzen-y replied:
That makes sense. Thank you for confirming that!

tenzen-y reviewed (commented):
Otherwise lgtm.
Will you work on conformance testings in a separate PR?

mimowo commented:
> Otherwise lgtm.

Thank you for reviewing.

> Will you work on conformance testings in a separate PR?

Separate PR was my original thought, but one PR also makes sense as the e2e tests are stable:
- "should mark indexes as failed when the FailIndex action is matched in podFailurePolicy" [link](https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=should%20mark%20indexes%20as%20failed%20when%20the%20FailIndex%20action%20is%20matched%20in%20podFailurePolicy); 
- "should execute all indexes despite some failing when using backoffLimitPerIndex" [link](https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=should%20execute%20all%20indexes%20despite%20some%20failing%20when%20using%20backoffLimitPerIndex)
- "should terminate job execution when the number of failed indexes exceeds maxFailedIndexes" [link](https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=should%20terminate%20job%20execution%20when%20the%20number%20of%20failed%20indexes%20exceeds%20maxFailedIndexes)

There are a couple of failures but all are for the Job `pull-kubernetes-e2e-gce-canary` and are accompanied by massive failures of other tests.

I pushed the commit: https://github.com/kubernetes/kubernetes/pull/130061/commits/66e45ff0db18f78c83a64332a5c35c53224d5bbe.

tenzen-y commented:
> Separate PR was my original thought, but one PR also makes sense as the e2e tests are stable:



> > Otherwise lgtm.
> 
> Thank you for reviewing.
> 
> > Will you work on conformance testings in a separate PR?
> 
> Separate PR was my original thought, but one PR also makes sense as the e2e tests are stable:
> 
> * "should mark indexes as failed when the FailIndex action is matched in podFailurePolicy" [link](https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=should%20mark%20indexes%20as%20failed%20when%20the%20FailIndex%20action%20is%20matched%20in%20podFailurePolicy)
> * "should execute all indexes despite some failing when using backoffLimitPerIndex" [link](https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=should%20execute%20all%20indexes%20despite%20some%20failing%20when%20using%20backoffLimitPerIndex)
> * "should terminate job execution when the number of failed indexes exceeds maxFailedIndexes" [link](https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=should%20terminate%20job%20execution%20when%20the%20number%20of%20failed%20indexes%20exceeds%20maxFailedIndexes)
> 
> There are a couple of failures but all are for the Job `pull-kubernetes-e2e-gce-canary` and are accompanied by massive failures of other tests (at least 184 failed tests in this [run](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/129983/pull-kubernetes-e2e-gce-canary/1888687121561030656)), so I think the failures are not indicative of an issue with the e2e tests, but with the CI job.
> 
> I pushed the commit: [66e45ff](https://github.com/kubernetes/kubernetes/commit/66e45ff0db18f78c83a64332a5c35c53224d5bbe).

Thank you for summarizing those. Let me try to review conformance and e2e tesgings.

cici37 commented:
/remove-sig api-machinery

mimowo commented:
I believe all so-far comments are addressed. PTAL @tenzen-y @atiratree

soltysh commented:
/triage accepted
/priority important-longterm

atiratree commented on test/e2e/apps/job.go:
```diff
@@ -593,13 +593,14 @@ done`}
 		backoffLimitPerIndex=1, allowing for one pod recreation before marking that indexed failed.
 	*/
-	ginkgo.It("should execute all indexes despite some failing when using backoffLimitPerIndex", func(ctx context.Context) {
+	framework.ConformanceIt("should execute all indexes despite some failing when using backoffLimitPerIndex", func(ctx context.Context) {
```
Before we mark these as a conformance, shouldn't we also check that the pods have corresponding `batch.kubernetes.io/job-index-failure-count` annotation since it is part of the feature? Or alternatively introduce an additional test for this?

tenzen-y replied:
Good point. In that case, do we need to implement additional check in a separate PR before we make this as conformance. After feature freeze (before test freeze), we could promote those e2e to conformance so that we can prepare the verification period if new verifications work correctly.

mimowo replied:
The options I can see:
1. defer promotion for later (separate PR) and extend the asserts to check the annotation, maybe we could do it still in 1.33, but not sure
2. promote the tests as they are, and later extend the asserts with the annotation
3. don't assert on the annotation as it is more of an implementation detail rather than API surface we want users to rely on

My preference order is `(3.) > (2.) > (1.)`.

My motivation to prefer (3.) is that it is just a bi-product of the mechanism. Similarly we don't have asserts in conformace tests for other things which are formally parts of the API, but we consider them as internal (just example which come to my head, but there are likely more):
- `status.uncountedTerminatedPods`
- labels matching the generated selector (see [here](https://github.com/kubernetes/kubernetes/blob/7fc8a86381d874e3bb47b6d343c20f93ace42981/pkg/registry/batch/job/strategy.go#L226)).

If we choose (2.) I would suggest just having a dedicated e2e added for them. We can consider its promotion independently, or even never promote.

But I'm open to other suggestions. cc @tenzen-y @atiratree @soltysh .

atiratree replied:
> My motivation to prefer (3.) is that it is just a bi-product of the mechanism. Similarly we don't have asserts in conformace tests for other things which are formally parts of the API, but we consider them as internal (just example which come to my head, but there are likely more):

Even if it is not part of the main API, it is mentioned there. Annotations are part of the API contract. Although we have the unit/integration for this, conformance E2Es serve a different purpose. For example, it shows what a job controller should implement. Apart from just forking the KCM, we offer a `managedBy` field so we should be more careful IMO.

I think we do not need a complex test, just a simple check at the end that the annotations have the expected value. IMO it would be best to go with (1.), but will give space for others to give an opinion.

mimowo replied:
> just a simple check at the end that the annotations have the expected value.

We need to be careful there too, at the end the Pods might be GCed so we may run into flakes. The easiest would be to check the annotation value before actually completing the Jobs.

mimowo replied:
>  Apart from just forking the KCM, we offer a managedBy field so we should be more careful IMO.

Well the additional assert will not help to verify user controllers in the context of managedBy, because there is no validation in k8s for the annotations as they are on Pods. We only strengthen the validation of Job status.

Also, I'm thinking the asserts will be tricky and can make the tests either flaky or less useful. Less consider this [test](https://github.com/kubernetes/kubernetes/blob/ad4d9b2f5119f2bcf742239018a397d3794cb6e0/test/e2e/apps/job.go#L602-L627) with the fragment:
```
		job, err := e2ejob.CreateJob(ctx, f.ClientSet, f.Namespace.Name, job)
		framework.ExpectNoError(err, "failed to create job in namespace: %s", f.Namespace.Name)

		ginkgo.By("Awaiting for the job to fail as there are failed indexes")
		err = e2ejob.WaitForJobFailed(ctx, f.ClientSet, f.Namespace.Name, job.Name)
		framework.ExpectNoError(err, "failed to ensure job completion in namespace: %s", f.Namespace.Name)

		ginkgo.By("Verifying the Job status fields to ensure all indexes were executed")
		job, err = e2ejob.GetJob(ctx, f.ClientSet, f.Namespace.Name, job.Name)
```
First, injecting the assert after WaitForJobFailed risks the Pods to be GCed, and it may depend on the CI job configuration with regards to GC configuration. Maybe we could check failure-count=1 or the Pod is Gone, but IMO it makes the test code hard to reason about. 

Second, between `CreateJob` and `WaitForJobFailed` 2 pods get created per index. So we have no guarantee which one we check. So, to make them stable we would need to split the process into 2 distinct phases, and this requires using different images which handle SIGTERM and after each step check the annotations. However, the phases will reduce the usefulness of the test, because part of the motivation is to detect race conditions in the process. By waiting for the replacements in the test code I think we reduce the chance of detecting such race conditions.

So, I think if we really want to test the annotations they deserve a dedicated test which is well thought to test them specifically. 

WDYT @atiratree ?

mimowo replied:
right, I didn't think of watches, let me re-think.

 Maybe indeed we could track pods in the watches and just check that per index we observed two of them with the annotation values. Still it will add complexity to the tests because we need to initialize the watches and assert on them in two entry points, so maybe indeed better to offload the current tests to make them more readable. The E2e tests also play a role of demonstrating to users how to use API so the extra code could complicate readability of them, but I get your point.

mimowo replied:
Ok, I started to write the asserts on annotations and it makes the test really beefy: https://gist.github.com/mimowo/df95ca0a4ad149f982da244e247c5509 - this is not done yet, I have the watch to track state of the Pods, but yet without asserts.

Still, I'm not sure this is the regime of e2e testing, as I mentioned before I don't see any difference between the annotation an other API surface-level things we only test in integration and unit tests:
- status.uncountedTerminatedPods
- Job's Pod finalizers
- Pod labels (UID and JobName) for selectors

The argument about the managedBy or forking controllers is also not clear to me. Say someone writes a 3party controller and instead of using annotations is saving the failure count in an external database (say mongodb). Is this a problem for the core k8s? I would say using annotations is just how k8s solves this problem, but not the only way.

mimowo replied:
wdyt @soltysh? 

@atiratree 
- can you elaborate why to cover this new annotation while the other above mentioned aspects of the API are not menioned. 
- seeing the draft of the test, would you be ok, to keep the test for annotations as a dedicated one which is not a conformance test?

soltysh replied:
> wdyt @soltysh?

Having that particular annotation tested is nice to have, but not a strict requirement. Your proposed order (merge now and expand later) seems reasonable to me.

atiratree replied:
> can you elaborate why to cover this new annotation while the other above mentioned aspects of the API are not menioned.

- we try to raise the testing bar over time and improve/test things we didn't test before https://groups.google.com/a/kubernetes.io/g/dev/c/6F3h0Z1QzVg/m/43tdYs_lAAAJ
- with the introduction with the `managedBy` field we have strengthened the API validation, but I think it is also good to continue with this on other fronts  

> seeing the draft of the test, would you be ok, to keep the test for annotations as a dedicated one which is not a conformance test?

> Having that particular annotation tested is nice to have, but not a strict requirement. Your proposed order (merge now and expand later) seems reasonable to me.

We can add the test later, but IMO it would be good to mark it as conformance once we are sure it is stable. As it can be used by other controllers to certify their behavior: https://kubernetes.io/docs/concepts/workloads/controllers/job/#delegation-of-managing-a-job-object-to-external-controller

mimowo replied:
I will try to follow up with completing this e2e test. Let's discuss its potential status as conformance in the future.

mimowo replied:
PR with the e2e test: https://github.com/kubernetes/kubernetes/pull/130390. PTAL.

atiratree reviewed (commented):
lgtm overall

tenzen-y reviewed (commented):
From my PoV, LGTM!
Thanks!

soltysh reviewed (approved):
/lgtm
/approve

soltysh commented:
Even though the changes are minor this has to go through api review, so tagging accordingly.

/label api-review

deads2k commented on test/integration/job/job_test.go:
```diff
@@ -1025,89 +1030,6 @@ func TestBackoffLimitPerIndex_DelayedPodDeletion(t *testing.T) {
 
-// TestBackoffLimitPerIndex_Reenabling tests handling of pod failures when
-// reenabling the BackoffLimitPerIndex feature.
-func TestBackoffLimitPerIndex_Reenabling(t *testing.T) {
```
@siyuanfoundation @jpbetz is there  way to set emulation version for a particular integration test to ensure this keeps working?

siyuanfoundation replied:
Yes, please follow the instructions in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#disablement-tests

mimowo replied:
Thanks! It works, re-added the test in the latest commit.

mimowo committed:
Promote e2e to conformance tests for JobBackoffLimitPerIndex

mimowo commented:
Rebased after https://github.com/kubernetes/kubernetes/pull/130390 got merged

mimowo commented:
cc @atiratree @deads2k PTAL. 

I have reverted the deletion of the TestBackoffLimitPerIndex_Reenabling test, and rebased after https://github.com/kubernetes/kubernetes/pull/130390.

atiratree commented on test/integration/job/job_test.go:
```diff
@@ -1068,6 +1073,7 @@ func TestBackoffLimitPerIndex_Reenabling(t *testing.T) {
 	validateIndexedJobPods(ctx, t, clientSet, jobObj, sets.New(1, 2), "", ptr.To("0"))
 
 	// Disable the feature
+	featuregatetesting.SetFeatureGateEmulationVersionDuringTest(t, feature.DefaultFeatureGate, utilversion.MustParse("1.32"))
```
Even though this works, I think it would be nicer to do this at the beginning of the test.

mimowo replied:
Sure, done.

atiratree commented:
Thx!
/lgtm

liggitt commented:
/approve
for API doc update

mimowo commented:
/assign @dims 
Could you review / approve the conformance tests?

dims commented:
@mimowo looking at the aggregated failures for these 3 tests:
- https://storage.googleapis.com/k8s-triage/index.html?test=Job%20should%20execute
- https://storage.googleapis.com/k8s-triage/index.html?test=Mark%20indexes%20as%20failed
- https://storage.googleapis.com/k8s-triage/index.html?test=Terminate%20job%20execution

There are some one-off spikes, can you please document if they were infra related or flakes? we would more more worried about promoting flakes to conformance, so we gotta dig just a bit deeper into these failures.

mimowo commented:
> There are some one-off spikes, can you please document if they were infra related or flakes? we would more more worried about promoting flakes to conformance, so we gotta dig just a bit deeper into these failures.

Sure, I did such an investigation last week and summarized here: https://github.com/kubernetes/kubernetes/pull/130061#issuecomment-2662525387. 

Checking the latest failures, this is the aggregated list of failed builds:
1. ci-kubernetes-kind-e2e-parallel-1-32: https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-kind-e2e-parallel-1-32/1894906346046754816 (328/6626 tests failed)
2. ci-kubernetes-e2e-ubuntu-ec2-arm64-containerd: https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-e2e-ubuntu-ec2-arm64-containerd/1891527770572328960 (404/6684 tests failed)
3. ci-kubernetes-e2e-ubuntu-ec2-arm64-containerd: https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-e2e-ubuntu-ec2-arm64-containerd/1892434032252162048 (408/6689 tests failed)

Digging into artifact logs:
1. failed in BeforeEach like most other failed tests with:
```
 I0227 01:29:09.026625 70939 util.go:360] Unexpected error while creating namespace: Post "https://127.0.0.1:33929/api/v1/namespaces": dial tcp 127.0.0.1:33929: connect: connection refused
```
So looks like API server was down and the test could not create the namespace.

2. the Job actually starts, but it does not make progress, I assume the Pods didn't schedule, because the DeferCleanup logs repeatedly (also for other tests):
```
{ failed [FAILED] checking for ready nodes: Not ready nodes: ", ip-172-31-2-94.ec2.internal"
In [DeferCleanup (Each)] at: k8s.io/kubernetes/test/e2e/framework/node/init/init.go:35 @ 02/17/25 19:50:53.446
```
and
```
I0217 19:43:53.445070 96562 helper.go:125] Waiting up to 7m0s for all (but 0) nodes to be ready
I0217 19:43:53.469734 96562 resource.go:131] Condition Ready of node ip-172-31-2-94.ec2.internal is false, but Node is tainted by NodeController with [{node.kubernetes.io/unreachable  NoSchedule 2025-02-17 16:59:03 +0000 UTC} {node.kubernetes.io/unreachable  NoExecute 2025-02-17 16:59:08 +0000 UTC}]. Failure
```
3. Similarly as in (2.) the Job starts, but it times out, and we see analogous logs indicating the nodes are unschedulable:
```
I0220 06:54:29.615870 98935 helper.go:125] Waiting up to 7m0s for all (but 0) nodes to be ready
I0220 06:54:29.640631 98935 resource.go:131] Condition Ready of node ip-172-31-14-97.ec2.internal is false, but Node is tainted by NodeController with [{node.kubernetes.io/unreachable  NoSchedule 2025-02-20 05:33:33 +0000 UTC} {node.kubernetes.io/unreachable  NoExecute 2025-02-20 05:33:43 +0000 UTC}]. Failure
```

So, I think all 3 failed builds can be categorized as infra failures.

dims commented:
/approve
/lgtm

thanks for documenting the failures @mimowo happy for us to land these conformance updates.
